### PR TITLE
TRT-274:   Support Historical Sippy Views

### PR DESCRIPTION
**Pull request by neisw**
[TRT-274](https://issues.redhat.com//browse/TRT-274): Add support for setting a fixed date to show results.

### CODE REVIEW

File: pkg/api/job_runs.go [hunk 1]
```diff
@@ -44,14 +45,19 @@ func (runs apiRunResults) limit(req *http.Request) apiRunResults {
...
 	table := "prow_job_runs_report_matview"
 	q, err := filter.FilterableDBResult(dbc.DB.Table(table), filterOpts, apitype.JobRun{})
 	if err != nil {
 		return nil, err
 	}
-	q = q.Where("release = ?", release)
+
+	if len(release) > 0 {
+		q = q.Where("release = ?", release)
+	}
```
Comment (stbenjam) [inline pkg/api/job_runs.go:58]: Would you mind moving this to a separate PR? I would rather land just this right away.

File: pkg/db/db.go [hunk 2]
```diff
@@ -28,7 +29,7 @@ type DB struct {
 	BatchSize int
 }
 
-func New(dsn string) (*DB, error) {
+func New(dsn string, reportEnd *time.Time) (*DB, error) {
```
Comment (stbenjam) [inline pkg/db/db.go:32]: I think it'd be cleaner if we separated the concerns of creating a DB client, and handling migrations in a Migrate() function.  You don't have to do it since this is a pre-existing thing but it would be nicer I think.

File: pkg/db/matviews.go [hunk 3]
```diff
@@ -3,29 +3,34 @@ package db
 import (
 	"fmt"
 	"strings"
+	"time"
 
 	"gorm.io/gorm"
 )
 
+const ReplaceTimeNow = "|||TIMENOW|||"
+const TimestampFormat = "2006-01-02 15:04:05"
```
Comment (stbenjam) [inline pkg/db/matviews.go:12]: Are these only used here? They should start with a lower case character, otherwise they're exported

File: pkg/db/query/payload_queries.go [hunk 4]
```diff
@@ -87,7 +89,7 @@ func GetLastPayloadStatus(db *gorm.DB, architecture, stream, release string) (st
 				FROM
 					release_tags
 				WHERE
-					architecture = ? AND stream = ? AND release = ?
+					architecture = ? AND stream = ? AND release = ? AND release_time < ?
```
Comment (stbenjam) [inline pkg/db/query/payload_queries.go:92]: The matviews do `<=` and `>=`, should these all be inclusive too?

File: sippy-ng/src/App.js [hunk 5]
```diff
@@ -120,26 +120,38 @@ export default function App(props) {
...
   const [fetchError, setFetchError] = React.useState('')
 
   const fetchData = () => {
     Promise.all([
       fetch(process.env.REACT_APP_API_URL + '/api/releases'),
       fetch(process.env.REACT_APP_API_URL + '/api/capabilities'),
+      fetch(process.env.REACT_APP_API_URL + '/api/report_date'),
     ])
-      .then(([releases, capabilities]) => {
+      .then(([releases, capabilities, reportDate]) => {
```
Comment (stbenjam) [inline sippy-ng/src/App.js:132]: Instead of wiring this in many places, I think it might be more appropriate to use a context and wrap the whole app like Capabilities.  What do you think?

**Review by stbenjam — COMMENTED** [review]
A few comments but looks pretty good.

File: pkg/db/db.go (hunk 2 above)
Comment (neisw) [reply pkg/db/db.go:32]: I'm definitely open to some refactoring to consolidate / cleanup.  I could create a story about this and the utility methods for start, end and boundary we touched on in the previous pr that we closed without merging.  Would prefer to get this in an agreeable state first though then refactor after merging if ok.

File: pkg/db/matviews.go (hunk 3 above)
Comment (neisw) [reply pkg/db/matviews.go]: yep, thanks!

File: sippy-ng/src/App.js (hunk 5 above)
Comment (neisw) [reply sippy-ng/src/App.js:132]: Agree, thanks for the pointer.  Let me review & refactor.
